docs/source/plugins/for_plugin_developers.rst

```diff
@@ -0,0 +1,274 @@
+blocks with their own GUIs that show up next to the main napari window. Rather,
+plugin developers must decide which of the current *hook specifications*
+defined by napari that they would like to implement.
+
+For a complete list of *hook specifications* that developers can implement, see
+the :ref:`hook-specifications`.
+
+A single plugin package may implement more than one *hook specification*, but
+may not declare more the one *hook implementation* for any given specification.
+
+
+.. NOTE::
+   One of the primary ways that we will extend the functionality of napari over
+   time is by identifying new ideas for *hook specifications* that developers
+   can implement.  If you have a plugin idea that requires napari to create a
+   new hook specification, we'd love to hear about it!  Please think about what
+   the signature of your proposed hook specification would look like, and where
+   within the napari codebase you'd like your hook implementation to be called,
+   and `open a feature request
+   <https://github.com/napari/napari/issues/new?template=feature_request.md>`_
+   in the napari issue tracker with your proposal.
+
+Let's take the :func:`~napari.plugins.hook_specifications.napari_get_reader`
+hook (our primary "reader plugin" hook) as an example.  It is defined as:
+
+.. code-block:: python
+
+   LayerData = Union[Tuple[Any], Tuple[Any, Dict], Tuple[Any, Dict, str]]
+   ReaderFunction = Callable[[str], List[LayerData]]
+
+   @napari_hook_specification(firstresult=True)
+   def napari_get_reader(
+       path: Union[str, List[str]]
+   ) -> Optional[ReaderFunction]:
+       ...
+
+Note that it takes a ``str`` or a ``list`` of ``str`` and either returns
+``None`` or a function.  From the docstring of the hook specification, we see
+that the implementation should return ``None`` if the path is of an
+unrecognized format, otherwise it should return a ``ReaderFunction``, which is
+a function that takes a ``str`` (the filepath to read) and returns a ``list``
+of ``LayerData``, where ``LayerData`` is any one of ``(data,)``, ``(data,
+meta)``, or ``(data, meta, layer_type)``.
+
+That seems like a bit of a mouthful!  But it's a precise (though flexible)
+contract that we can follow, and know that napari will handle the rest.
+
+
+Step 2: Write your hook implementation
+--------------------------------------
+
+Once you have identified the :ref:`hook specification <hook-specifications>`
+that you want to implement, you have to create a *hook implementation*: a
+function that accepts the arguments specified by the hook specification
+signature and returns a value with the expected return type.
+
+Here's an example hook implementation for
+:func:`~napari.plugins.hook_specifications.napari_get_reader` that enables
+napari to open an imaginary ``.ext`` filetype.
+
+.. code-block:: python
+
+   from pluggy import HookimplMarker
+
+   # we'll get to this line and the decorator below in just a minute
```

Comment:
maybe specify where this is explained?

Reply:
And, seeing that the code requires explanation, I wonder whether we should show importing `napari_hook_implementation` here, which is more self-explanatory, even if only on a superficial level. Then, later where it is explained, you can say, "and, bonus, you don't actually need to import napari to create a plugin! tada!" and explain the fancier pluggy version. What do you think? (Also @sofroniewn)